staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/handler.go

```diff
@@ -265,17 +351,34 @@ func (rdm *resourceDiscoveryManager) calculateAPIGroupsLocked() []apidiscoveryv2
 	regenerationCounter.Inc()
 	// Re-order the apiGroups by their priority.
 	groups := []apidiscoveryv2beta1.APIGroupDiscovery{}
-	for _, group := range rdm.apiGroups {
+
+	groupsToUse := map[string]apidiscoveryv2beta1.APIGroupDiscovery{}
+	sourcesUsed := map[string]string{}
+
+	for key, group := range rdm.apiGroups {
+		// Skip groups we've already seen before. Only DefaultSource
+		// takes precedence
+		if _, seen := groupsToUse[key.name]; seen && key.source != DefaultSource {
```

Comment:
I believe if you have two non-default sources with the same group, this code will use a random source, right? I think it should be deterministic. If source isn't default, I think you should compare key.source with the previously seen source and pick the lexicographically least.